pkg/loki/query.go

```diff
@@ -0,0 +1,222 @@
+	}
+	switch key {
+	case startTimeKey:
+		q.addURLParam(startParam, value)
+	case endTimeTimeKey:
+		q.addURLParam(endParam, value)
+	case timeRangeKey:
+		r, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return err
+		}
+		q.addURLParam(startParam, strconv.FormatInt(time.Now().Unix()-r, 10))
+	case limitKey:
+		q.addURLParam(limitParam, value)
+	// Attributes that have a special meaning and need to be treated apart
+	case matchParam, flowDirParam:
+		q.specialAttrs[key] = value
+	// Stream selector labels
+	case "SrcNamespace", "SrcWorkload", "DstNamespace", "DstWorkload":
+		q.processStreamSelector(key, strings.Split(value, ","))
+	// IP filter labels
+	case "DstAddr", "SrcAddr", "DstHostIP", "SrcHostIP":
+		q.processIPFilters(key, strings.Split(value, ","))
+	default:
+		return q.processLineFilters(key, strings.Split(value, ","))
+	}
+	return nil
+}
+
+// PrepareToSubmit returns a new Query that already handles the special behavior of some attributes
+// that mustn't be used as part of a generic query.
+func (q *Query) PrepareToSubmit() (*Query, error) {
+	var out *Query
+	// If match=any, it converts the query to a query that matches when only of the given
+	// attributes match
+	if match := q.specialAttrs[matchParam]; match == anyMatchValue {
+		out = q.convertToAnyMatch()
+	} else {
+		// copy receiver query
+		cp := *q
+		out = &cp
+	}
+
+	// Append app stream selector, which will apply whichever matching criteria (any or all)
+	out.streamSelector = append(out.streamSelector,
+		stringLabelFilter("app", labelEqual, "netobserv-flowcollector"))
+
+	// Filter by flow direction independently of the matching criteria (any or all)
+	if flowDir, ok := out.specialAttrs[flowDirParam]; ok {
+		if err := out.processLineFilters(flowDirParam, []string{flowDir}); err != nil {
+			return nil, err
+		}
+	}
+	return out, nil
+}
+
+func (q *Query) addURLParam(key, val string) {
+	q.urlParams = append(q.urlParams, [2]string{key, val})
+}
+
+func (q *Query) processStreamSelector(key string, values []string) {
+	regexStr := strings.Builder{}
+	for i, value := range values {
+		if i > 0 {
+			regexStr.WriteByte('|')
+		}
+		//match any caracter before / after value : .*VALUE.*
+		regexStr.WriteString(".*")
+		regexStr.WriteString(value)
+		regexStr.WriteString(".*")
+	}
+
+	if regexStr.Len() > 0 {
+		q.streamSelector = append(q.streamSelector,
+			stringLabelFilter(key, labelMatches, regexStr.String()))
+	}
+}
+
+// filterIPInLine assumes that we are searching for that IP addresses as part
+// of the log line (not in the stream selector labels)
+func (q *Query) processIPFilters(key string, values []string) {
+	for _, value := range values {
+		q.labelFilters = append(q.labelFilters, ipLabelFilter(key, value))
+	}
+}
+
+func (q *Query) processLineFilters(key string, values []string) error {
+	regexStr := strings.Builder{}
+	for i, value := range values {
+		if i > 0 {
+			regexStr.WriteByte('|')
```

Comment:
This makes always an 'OR' between multiples values in the same `key` filter.
I would expect to have a 'AND' here in case of 'Match all' query option.

![image](https://user-images.githubusercontent.com/91894519/152980832-76e9b7d9-fcf2-4463-a215-627284e277ac.png)

What do you think ?

Reply:
This is the behavior that we had already before this PR. I guess it anyway does "or" because a pod can't be "net" and "zzz" at the same time.

I can modify it so e.g. the Pod "contains" net AND zzz at the same time but we all need to be aware that this is a breaking change.

Reply:
Yes I thought this PR was about to change the current behavior for both cases: 
- Match all = 'AND' between values whatever if it's from same / different filters (breaking change)
- Match any = 'OR' current behavior between values in the same filter

Let's ask to everybody before making any changes